### PR TITLE
[NFC] Refactor Pattern Binding Application in Result Builders

### PR DESCRIPTION
**Pull request by CodaFi**
Remove the duplication between this code and CSApply by targeting the pattern binding decl instead of all of its entries.

### CODE REVIEW

File: lib/Sema/BuilderTransform.cpp [hunk 1]
```diff
@@ -1238,9 +1213,11 @@ class BuilderClosureRewriter
 
       // Handle pattern bindings.
       if (auto patternBinding = dyn_cast<PatternBindingDecl>(decl)) {
-        finishPatternBindingDecl(patternBinding);
-        TypeChecker::typeCheckDecl(decl);
-        newElements.push_back(decl);
+        auto resultTarget = rewriteTarget(SolutionApplicationTarget{patternBinding});
+        if (!resultTarget)
+          continue;
```
Comment (CodaFi) [inline lib/Sema/BuilderTransform.cpp:1218]: @xedin Do you think it's okay that we drop broken rewrite targets on the floor? I can restore the old behavior by inserting `patternBinding` into the element buffer instead.

Comment (CodaFi) [inline lib/Sema/BuilderTransform.cpp:1218]: @xedin Do you think it's okay that we drop broken rewrite targets on the floor? I can restore the old behavior by inserting `patternBinding` into the element buffer instead.

Comment (CodaFi) [reply lib/Sema/BuilderTransform.cpp:1218]: Perhaps this should assert that the result is never null and push on. Doesn’t look like CSApply returns None here.

Comment (xedin) [reply lib/Sema/BuilderTransform.cpp:1218]: Yeah, I think we need two asserts here - first one for getting a solution application target from constraint system (similar to what old code did for individual patterns) to make sure that it exists, and second - that rewrite succeeded.

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
⛵

**Pull request merged**